dotnet/src/IntegrationTests/Connectors/Oobabooga/OobaboogaTextCompletionTests.cs

```diff
@@ -0,0 +1,72 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
```

Comment:
Could you please update this [README.md](https://github.com/microsoft/semantic-kernel/blob/main/dotnet/src/IntegrationTests/README.md) and add instructions how to run tests locally? We have a plan to do the same for HuggingFace later. 

Also, since these are integration tests which require local setup and can be verified only manually, it would be great if we will add unit tests for this functionality. You can use [HuggingFace unit tests](https://github.com/microsoft/semantic-kernel/blob/main/dotnet/src/Connectors/Connectors.UnitTests/HuggingFace/TextCompletion/HuggingFaceTextCompletionTests.cs) as an example.

Reply:
For instructions, see [#6d1a4c5840ae2fb7c24a58fdbcc6311e2610bc00](https://github.com/microsoft/semantic-kernel/pull/1357/commits/6d1a4c5840ae2fb7c24a58fdbcc6311e2610bc00). 
I kept them minimal, since I didn't know how much details your wanted to add to that Readme, but overall oobabooga install is pretty straight forward (installing all the accelerators and optimizers is another matter, and may require to play with custom wheels, cuda compilation switching to wsl2, depending on your GPUs hardware, environment and target models, but I'm pretty sure this is temporary). Anyway if you have issues, let me know. If you have a decent GPU, you can go a long way with the default install, and the gradio web ui has tons of parameters that you can hot swap (in the flavor of stable-diffusion web ui as officially stated).

Reply:
Unit tests were hard.
First and foremost, why isn't System.Linq.Async Nuget package used throughout the solution? It seems to me the current situation with half baked replacements is suboptimal, and requires pragmas to get through.
Then my main issue was that you can't get those mocks as easily with ClientWebsocket as you do with Httpclient (handler is internal). I ended up introducing an actual test websocket server that registers on a local port. It works well locally, but hopefully this won't break depending on the test running environment.
Eventually, I believe this is a useful addition since that test server will allow more complex tests than simple mocks would have allowed.

I also updated the integration tests, but you might want to change or complete them according to your needs. 
Hopefully all of this is enough for you to get started and to take over though...

Reply:
Also I guess there is a bit of overlap/redundancy with HuggingFace connector now, and ultimately, you'd want to abstract and mutualise some of the code I introduced.

Reply:
Thanks @jsboige , this is awesome! In case `ClientWebsocket` is not mockable, let's remove unit tests for it and leave that responsibility for integration tests. Let's use `HttpClient` only for unit tests.

Reply:
> Thanks @jsboige , this is awesome! In case `ClientWebsocket` is not mockable, let's remove unit tests for it and leave that responsibility for integration tests. Let's use `HttpClient` only for unit tests.

Well, I ended up mocking the websocket server, which works great on my environment, so unless there is anything preventing it to run on the test server, I'd say we're good to go and that is still a good addition, which will allows for subtle fine-grained unit test since we have complete control on the way the server handles the incoming byte packets.

For instance, [that test](https://github.com/microsoft/semantic-kernel/blob/7025f2442b2cbe0ae9ab8bb1dc70d9525a19ce9c/dotnet/src/Connectors/Connectors.UnitTests/Oobabooga/TextCompletion/OobaboogaTextCompletionTests.cs#L125) aggregates the incoming chunks on the server side and checks that the request was properly submitted, whereas [this one](https://github.com/microsoft/semantic-kernel/blob/7025f2442b2cbe0ae9ab8bb1dc70d9525a19ce9c/dotnet/src/Connectors/Connectors.UnitTests/Oobabooga/TextCompletion/OobaboogaTextCompletionTests.cs#L152) has the server send back the expected response chunks and checks that the client properly handles it.